syntax = "proto3";

// The greeting service definition.
// python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. mat.proto
service Mat {
  // Sends a greeting
  rpc login (LoginRequest) returns (UserLoginResponse);
  rpc GameStream (stream ActionRequest) returns (stream ActionResponse);

}


//////////////////////////////////////////////////////////////////////////////
// Request Messsages
//////////////////////////////////////////////////////////////////////////////

// A request with just the user's credentials.
message LoginRequest {
    string user_id = 1;   //用户名
    string user_pin = 2;  //登录密码
}

message ActionRequest {
    string user_id = 1;       //用户名
    string user_pin = 2;      //登录密码

    enum RequestType {
      HeartBeat = 0;            //client心跳
      RequestReady = 1;         //client已准备
      RequestAsk = 2;           //client申请相应题号的密钥
      RequestAnswer = 3;        //client
    }

    RequestType msg_type = 3; //命令类型
    //int32 question_num = 4;   //申请的题号
    string game_info = 4;     //游戏信息
}

//////////////////////////////////////////////////////////////////////////////
// Response Messsages
//////////////////////////////////////////////////////////////////////////////
message UserLoginResponse {
    string user_id = 1;
    string user_pin = 2;
    double init_capital = 3; //玩家初始积分
    bool success = 4;        //是否登录成功
    string reason = 5;       //原因
}

message ActionResponse {
    string user_id = 1;
    string user_pin = 2;

    enum RequestType {
      StateInit = 0;       //client决
      ResponseAsk = 1;     //返回题号与密钥
      ResponseAnswer = 2;  //返回题号与准确率
      ResponseInfo = 3;    //返回通知信息
    }

    RequestType msg_type = 3;     //命令类型
    string game_info = 4;        //游戏信息
}